Intersection of two wires and related utilities

PetrilloAtWork opened the pull request:
This is the proposed resolution of [Redmine issue #25977](https://cdcvs.fnal.gov/redmine/issues/25977).

Discussion and description of the design can be found in that ticket.
As usual, the core new features are unit-tested, and all the new functions and methods are Doxygen-documented.
In particular, unit testing covers the lowest level free functions (intersection of lines) and the top level (`geo::GeometryCore` methods).

In addition, a commit is updating the Doxygen format of `geo::WireGeo`, which was somehow broken with methods escaping their documentation group.

PetrilloAtWork committed:
Added `geo::GeometryCore::WiresIntersect()` for unchecked wire intersection

This is a candidate resolution for Redmine issue #25977, with option 1.1
(`GeometryCore` const member function).

PetrilloAtWork committed:
Added intersection functions for geo::WireGeo.

This is the extended solution of Redmine issue #25977.

PetrilloAtWork committed:
Format update for geo::WireGeo documentation

FNALbuild commented:
A new Pull Request was created by  @PetrilloAtWork (Gianluca Petrillo) for develop.

It involves the following packages:

larcorealg


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//957/console

PetrilloAtWork committed:
Added unit test of intermediate wire intersection functions

Since I realised it would be cheap.

FNALbuild commented:
Pull request #17 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//958/console

knoepfel commented on larcorealg/Geometry/LineClosestPoint.tcc:
```diff
@@ -0,0 +1,129 @@
+
+  // aliases for quick notation
+  Point const& c1 = startA;
+  Vector const& w1 = dirA;
```
Minor cleanup:

```c++
auto const& [c1, w1] = std::tie(startA, dirA);
```

PetrilloAtWork replied:
I am not sure readability benefits from it, but ok, I'll pay the price of including `tuple`.

BTW: you can use the "suggestion" feature of GitHub for this type of comments (I think it's enough to put `suggesion` after the opening triple back tick):
```suggestion
  auto const& [c1, w1] = std::tie(startA, dirA);
```
Why? because it looks cool.

knoepfel replied:
Yes, it's cool. 😃

knoepfel commented:
@PetrilloAtWork, do you have an example of how this code can/will be used?

It's hard to determine from the PR whether the additions are appropriate for the use case.  In particular, I'm not convinced that member functions are appropriate for providing this capability, and the `locOnLines` output argument does not appear to be useful (to me) in the case of non-unit vectors.

Seeing how this code would be used "in the wild" would help guide our review of it.  Thanks!

PetrilloAtWork commented:
One example is [vertex reconstruction code in `sbncode`](https://github.com/SBNSoftware/sbncode/blob/2e5e4ac8542d81039e8f4d0c1270daf2f6a5805c/sbncode/TPCReco/VertexStub/StubMergeAlgorithms.cxx#L156-L203): here the context is _art_-independent. Geometry is needed to translate the wire ID stored in the hit objects into actual physical location of the wires (and for nothing else), after which the intersection is performed (in some definitely-non-LArSoft way).
In `larreco` I see it used almost always starting from wire ID (usually from hits).

In addition to the wild, there is the explicit intention of having the math done in a single place to be used as implementation in `geo::GeometryCore` (which I failed the moment I split `LineClosestPoint()` in two variants).

Also consider that this commit follows from the discussion in the [issue](https://cdcvs.fnal.gov/redmine/issues/25977), where I was asking which of three approaches was better to take, and I was answered "go ahead". This is an "all of the above" implementation. 

To address your two points directly:

> I'm not convinced that member functions are appropriate for providing this capability

Member functions of `geo::GeometryCore` or of `geo::WireGeo`? I would argue for the member functions in `geo::GeometryCore`, since one of them is there already (arguments `geo::WireID`) while the other (arguments `geo::WireGeo`) is basically for free and it's faster if the ID → wire lookup has been already performed.
Intersection with another `geo::WireGeo` makes sense to me too, but I don't have particular feelings in seeing it gone.
Note however that the order of the operands does matter to some extent, because the returned point is located _on_ the first wire (`this` in case of the member function).

> the `locOnLines` output argument does not appear to be useful (to me) in the case of non-unit vectors

The functionality is there mostly because that additional information is useful to test for containment: that test is performed by `geo::GeometryCore::WireIDsIntesection()` (if I got the name right) and in that case it uses unit vectors. So the same functionality for non-unit vectors is not required. Several possible options:
1. leave it as it is (return value in unit of the vectors), which you are questioning;
2. change it to return the distance in the same unit as the other version, i.e. global units;
3. remove the additional output;
4. remove the function altogether and trust the users to do the normalization if needed.

I don't have, again, particular preferences, although I mildly tend not to remove the function altogether.

knoepfel commented on larcorealg/Geometry/GeometryCore.cxx:
```diff
@@ -1353,6 +1353,16 @@ namespace geo {
+  {
+    assert(!wire2.isParallelTo(wire1));
+    return geo::WiresIntersection(wire1, wire2);
+  } // GeometryCore::WiresIntersect()
```
As this function does not rely on `GeometryCore`'s state, it should not be a member function--yes, this breaks with "prior art", but I have a comment there, too.  The `geo::WiresIntersection` free function should be sufficient.  

Additionally, this member function appears to be a wrapping of the free function--however, there is a precondition that is checked here (i.e. the assert) that is already checked in the underlying implementation (`LineClosestPointWithUnitVectors`).

PetrilloAtWork replied:
> this function does not rely on GeometryCore's state

Practically, this is not true: all `geo::WireGeo` instances in LArSoft are owned by `geo::GeometryCore`. Does this change your statement?

knoepfel replied:
No, my statement is unchanged.  There is nothing in the `geo::WiresIntersection(a, b)` function that requires a `geo::GeometryCore` object--everything is passed in as function arguments and no `this` pointer is required.  The intersection function does not care where those `geo::WireGeo` objects came from nor who owns them.  And that's good--it's easier to test things and demonstrates a better separation of concerns.

knoepfel commented on larcorealg/Geometry/GeometryCore.cxx:
```diff
@@ -1353,6 +1353,16 @@ namespace geo {
+
+
   //......................................................................
   bool GeometryCore::WireIDsIntersect(
```
Many of `GeometryCore`'s member functions appear to be merely "convenience" functions that do not actually use `GeometryCore`'s internal state (such as `WireIDsIntersect`).  This creates unnecessary entanglements.  The `GeometryCore` class should gradually be pruned of its "add-ons", so that only member functions that require access to data members are provided.

Tackling that problem was not in the scope of this PR, but it motivates the reason for not supporting a `geo::WiresIntersection` member function.

knoepfel commented on larcorealg/Geometry/LineClosestPoint.tcc:
```diff
@@ -0,0 +1,129 @@
+  double const inv_den = 1.0 / (cet::square(w1w2) - (w1w1 * w2w2));
+  double const t = ((dcw2 * w1w2) - (dcw1 * w2w2)) * inv_den;
+
+  if (locOnLines) {
```
This is a bit too subtle.  Based on the interface, the user might expect the following call pattern:

```c++
std::pair<double, double>* locOnLines = nullptr;
auto pt = LineClosestPoint(startA, dirA, startB, dirB, locOnLines);
auto [t, u] = *locOnLines; // Boom!
delete locOnLines;
```

I think you're better off creating an overload set instead of providing a default value for the last argument:

```c++
auto pt1 = LineClosestPoint(startA, dirA, startB, dirB); // ........ // 4-arg. version
std::pair<double, double> locOnLines;
auto pt2 = LineClosestPoint(startA, dirA, startB, dirB, locOnLines); // 5-arg. version
assert(pt1 == pt2);
```

PetrilloAtWork replied:
Ok.
Although nobody should be expecting from LArSoft a C-flavoured interface like the one you describe.

knoepfel reviewed (commented):
@PetrilloAtWork, thank you for the response.  I've provided a few comments, which I hope will provide some perspective re. my original remarks.  The overall functionality you provide is excellent, we just want to find a maintainable solution, possibly breaking with previous coding patterns.

PetrilloAtWork committed:
Minor change from review of issue 25977

PetrilloAtWork committed:
Wire intersection utility interface split.

All information is produced by return value, and two different signatures are offered to users according to the desired information.
This implements a comment from the review of Issue 25977.

PetrilloAtWork committed:
Updated test to use BOOST_TEST()

FNALbuild commented:
Pull request #17 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//970/console

PetrilloAtWork committed:
Wire intersection functions now return a struct.

PetrilloAtWork committed:
Workaround to assign from intersection result struct with std::tie()

FNALbuild commented:
Pull request #17 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//971/console

PetrilloAtWork commented:
@knoepfel Thank you for your comments. I have left answers to them in the existing conversations.
I wait for an answer to one of them, while the other two can be resolved directly.

The first of the new commits implemented the first of the suggestions. Nothing has been done yet for the second conversation, pending an answer.
For the third one, I have implemented a split of interface. That is the second commit, plus the fourth and fifth: more on that below.
The last of the commits updates the unit test to the new `BOOST_TEST()`, since that is LArSoft's recommendation.

### Split interface

Each function or class method with an optional output parameter ("offsets") has been split into a version avoiding the additional output, and another one returning the offsets together with the intersection point. The name of the functions was intentionally chosen to be different in the two cases to reflect the different usage (this turns out also to be necessary when returning everything as return value of the functions).
The underlying implementation is still the same, since it was intended to skip operations not requested while keeping the math in a single spot.

I have chosen to return all the results as an aggregate return value. It has some advantages, like clearly separating input and output visually, and making some code more compact. It also has drawbacks, like the unusual `std::tie()` syntax.

I had to choose between returning a `tuple` or a dedicated data structure.
The `tuple` option has as drawbacks the unfriendly `std::get()` interface to access the elements, and above all the obscure data type: `std::tuple<geo::Point_t, double, double`: what are those `double`? how are they different? only documentation knows.
The dedicated `struct` has the advantage that its data members have meaningful names and their access is familiar to everybody; but it has the drawback that, while structure binding covers the initialisation of variables at definition time, as far as I know it is not possible to use that in assignments (not even a thing like `std::tie(point, ofsA, ofsB) = LineIntersectionWithOffsets(a, b);`).
There is a actually way to implement that (the last commit), but implicit type cast is not generally a great choice (this struct, though, is expected to be short-lived).

The second commit of the pull request implements the return value as `tuple`. The fourth implements it as a dedicated `struct`. The fifth introduces the workaround to have assignments to `std::tie()` work.

knoepfel commented:
@PetrilloAtWork, the updated PR is an improvement.  Thank you.  I still stand by my comment regarding not adding the `geo::GeometryCore::WiresIntersect(...)` member function.  However, I agree with your solution for the `locOnLines` issue.

PetrilloAtWork committed:
Removed geo::GeometryCore::WiresIntersection() method.

FNALbuild commented:
Pull request #17 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//972/console

PetrilloAtWork commented:
With the last commit, I believe I have addressed all the comments from the reviewers.
Let me know if instead more work is required.

knoepfel commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12049&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12050&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5666&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12050&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5914&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12050&builds=

FNALbuild commented:
-sbnd tests failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/6891&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12050&builds=

FNALbuild commented:
-icarus tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/2276&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12050&builds=

FNALbuild commented:
-dune tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/11658&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12050&builds=

FNALbuild commented:
-uboone tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6368&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12050&builds=

absolution1 commented:
Looks like SBND's CI is currently broken, so we can probably make do with a pinky promise in the mean time.
@PetrilloAtWork are any changes in the CI output expected?  I assume not...

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12066&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12067&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5670&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12067&builds=

FNALbuild commented:
+argoneut tests OK on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5918&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12067&builds=

FNALbuild commented:
-sbnd tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/6900&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12067&builds=

FNALbuild commented:
-icarus tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/2284&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12067&builds=

FNALbuild commented:
-dune tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/11662&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12067&builds=

lgarren commented:
@absolution1  the SBND CI tests were fixed

FNALbuild commented:
-uboone tests warning on slf7 for e20:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/6372&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/12067&builds=

PetrilloAtWork commented:
> @PetrilloAtWork are any changes in the CI output expected? I assume not...

I expect no changes. If that turns out not to be the case, it should be investigated.

lgarren commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests.
